File: lib/models/order.dart
import 'product.dart';

class Order {
  final String id;
  final String userId;
  final double totalAmount;
  final String status;
  final String paymentMethod;
  final String paymentStatus;
  final String shippingAddress;
  final String notes;
  final List<OrderItem> orderItems;
  final DateTime createdAt;

  Order({
    required this.id,
    required this.userId,
    required this.totalAmount,
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.shippingAddress,
    required this.notes,
    required this.orderItems,
    required this.createdAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    var itemsList = json['order_items'] as List? ?? [];
    List<OrderItem> items =
        itemsList.map((item) => OrderItem.fromJson(item)).toList();

    return Order(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      totalAmount: (json['total_amount'] ?? 0).toDouble(),
      status: json['status'] ?? '',
      paymentMethod: json['payment_method'] ?? '',
      paymentStatus: json['payment_status'] ?? '',
      shippingAddress: json['shipping_address'] ?? '',
      notes: json['notes'] ?? '',
      orderItems: items,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'total_amount': totalAmount,
      'status': status,
      'payment_method': paymentMethod,
      'payment_status': paymentStatus,
      'shipping_address': shippingAddress,
      'notes': notes,
      'order_items': orderItems.map((item) => item.toJson()).toList(),
      'created_at': createdAt.toIso8601String(),
    };
  }

  String get formattedTotal {
    return 'Rp ${totalAmount.toStringAsFixed(0).replaceAllMapped(
          RegExp(r'(\d{1,3})(?=(\d{3})+(?!\d))'),
          (Match m) => '${m[1]}.',
        )}';
  }

  String get statusDisplay {
    switch (status.toLowerCase()) {
      case 'pending':
        return 'Pending';
      case 'confirmed':
        return 'Confirmed';
      case 'shipped':
        return 'Shipped';
      case 'delivered':
        return 'Delivered';
      case 'cancelled':
        return 'Cancelled';
      default:
        return status;
    }
  }
}

class OrderItem {
  final String id;
  final String productId;
  final int quantity;
  final double price;
  final double subtotal;
  final Product product;

  OrderItem({
    required this.id,
    required this.productId,
    required this.quantity,
    required this.price,
    required this.subtotal,
    required this.product,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'] ?? '',
      productId: json['product_id'] ?? '',
      quantity: json['quantity'] ?? 1,
      price: (json['price'] ?? 0).toDouble(),
      subtotal: (json['subtotal'] ?? 0).toDouble(),
      product: Product.fromJson(json['product'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'product_id': productId,
      'quantity': quantity,
      'price': price,
      'subtotal': subtotal,
      'product': product.toJson(),
    };
  }
}

File: backend/models/order.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type Order struct {
	ID            string         `gorm:"type:char(36);primaryKey" json:"id"`
	UserID        string         `gorm:"type:char(36);not null;index" json:"user_id"`
	TotalAmount   float64        `gorm:"type:decimal(12,2);not null" json:"total_amount"`
	Status        string         `gorm:"type:varchar(20);default:'pending';index" json:"status"`
	PaymentMethod string         `gorm:"type:varchar(50)" json:"payment_method"`
	PaymentStatus string         `gorm:"type:varchar(20);default:'pending'" json:"payment_status"`
	ShippingAddr  string         `gorm:"type:text" json:"shipping_address"`
	Notes         string         `gorm:"type:text" json:"notes"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `gorm:"index" json:"-"`

	User       User        `gorm:"foreignKey:UserID" json:"user,omitempty"`
	OrderItems []OrderItem `gorm:"foreignKey:OrderID" json:"order_items,omitempty"`
}

func (Order) TableName() string {
	return "orders"
}

type OrderItem struct {
	ID        string         `gorm:"type:char(36);primaryKey" json:"id"`
	OrderID   string         `gorm:"type:char(36);not null;index" json:"order_id"`
	ProductID string         `gorm:"type:char(36);not null;index" json:"product_id"`
	Quantity  int            `gorm:"not null" json:"quantity"`
	Price     float64        `gorm:"type:decimal(12,2);not null" json:"price"`
	Subtotal  float64        `gorm:"type:decimal(12,2);not null" json:"subtotal"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Product Product `gorm:"foreignKey:ProductID" json:"product,omitempty"`
}

func (OrderItem) TableName() string {
	return "order_items"
}

type OrderResponse struct {
	ID            string              `json:"id"`
	UserID        string              `json:"user_id"`
	TotalAmount   float64             `json:"total_amount"`
	Status        string              `json:"status"`
	PaymentMethod string              `json:"payment_method"`
	PaymentStatus string              `json:"payment_status"`
	ShippingAddr  string              `json:"shipping_address"`
	Notes         string              `json:"notes"`
	OrderItems    []OrderItemResponse `json:"order_items"`
	CreatedAt     time.Time           `json:"created_at"`
}

type OrderItemResponse struct {
	ID        string          `json:"id"`
	ProductID string          `json:"product_id"`
	Quantity  int             `json:"quantity"`
	Price     float64         `json:"price"`
	Subtotal  float64         `json:"subtotal"`
	Product   ProductResponse `json:"product"`
}

func (o *Order) ToResponse() OrderResponse {
	items := make([]OrderItemResponse, len(o.OrderItems))
	for i, item := range o.OrderItems {
		items[i] = OrderItemResponse{
			ID:        item.ID,
			ProductID: item.ProductID,
			Quantity:  item.Quantity,
			Price:     item.Price,
			Subtotal:  item.Subtotal,
			Product:   item.Product.ToResponse(),
		}
	}

	return OrderResponse{
		ID:            o.ID,
		UserID:        o.UserID,
		TotalAmount:   o.TotalAmount,
		Status:        o.Status,
		PaymentMethod: o.PaymentMethod,
		PaymentStatus: o.PaymentStatus,
		ShippingAddr:  o.ShippingAddr,
		Notes:         o.Notes,
		OrderItems:    items,
		CreatedAt:     o.CreatedAt,
	}
}



